### ISSUE TITLE
Inline Javascript is not interpreted when Repeater is AJAX-loaded

### ISSUE THREAD

**Issue by Toutouwai** [processwire/processwire-issues#103]
### Short description of the issue

When inputfields are rendered in a Repeater item that is dynamically loaded (AJAX), any inline Javascript is not interpreted. This is a problem for inputfields that are customised with Javascript added in a render hook. Or any inputfield that might include Javascript in its render method.

### Expected behavior

Inline Javascript is interpreted for fields that are AJAX-loaded according to field visibility settings (Closed + load only when opened). Expectation is that AJAX-loaded Repeater items would be the same.

### Actual behavior

Inline Javascript is not interpreted.

### Steps to reproduce the issue

My case involves a module that appends Javascript to image fields to allow for easy insertion into CKEditor fields. But take a simple example hook:

```
$this->addHookAfter('InputfieldText::render', function($event) {
    $out = "<script>console.log('Field was rendered');</script>";
    $event->return .= $out;
});
```

This hook will log a message to the console when a text field is rendered normally. It also works when a text field is AJAX-loaded: the script is interpreted when the field loads and the message is logged. But the script is not interpreted when the field is rendered in an AJAX-loaded Repeater item. Not sure why this should be. 

### Setup/Environment

* ProcessWire version: 3.0.42

**Comment by ryancramerdesign** [comment]
I'm not aware of anything we are doing to remove inline scripts. But I have found that rendering inline scripts with ajax is rarely reliable. I'm not sure if it's something with browser security or jQuery or what, but it rarely seems to work. The alternative that I use is to have event listeners be ready ahead of time to look for the "reloaded" event on an Inputfied. Then perform the initialization with that. If you search the core codebase for "reloaded" you'll find lots of examples, but it basically works like this:
~~~~~
$(document).on('reloaded', '.Inputfield', function(event, source) {
  if(typeof source != "undefined") {
    // source is a string indicating what triggered the reloaded event
    // though chances are you won't need it
  }
  // perform any needed initialization code on $(this), which is an ".Inputfield" element
  console.log('Field was rendered: ' + $(this).children('.InputfieldHeader').text()); 
}); 
~~~~~
This listener would need to be established when the page originally renders. If you need to focus in on a specific Inputfield type, then you'd replace the ".Inputfield" in the first line with the Inputfield type, i.e. ".InputfieldDatetime".

**Comment by Toutouwai** [comment]
jQuery does mess with scripts in HTML that is ajax-loaded, but I'm not very experienced with ajax and it's difficult to find documentation on exactly how they are treated. The [docs for load()](http://api.jquery.com/load/#script-execution) say that scripts are executed so long as there is no selector expression - not sure if that is the same for all jQuery ajax methods or if it applies in the Repeater case. It also seems that script elements are parsed out of the HTML and rendered and the end of the loaded elements: https://www.bennadel.com/blog/1605-jquery-ajax-strips-script-tags-and-inserts-them-after-parent-most-elements.htm

The dataType setting would also have an effect. It's not explicitly set for Repeater ajax loads so I suppose the automatic inferring could be going wrong. If the dataType is HTML then scripts are supposed to execute:

> "html": Returns HTML as plain text; included script tags are evaluated when inserted in the DOM.

I'm aware of the 'reloaded' event, but the issue then is what to hook in order to insert the script that listens for the event. If ProcessPageEdit is hooked then the script is inserted for every page edit regardless of if it is needed. If Inputfield::renderReady were hookable (see [this request](https://github.com/processwire/processwire-requests/issues/12)) this may be a solution.

Also, adding inline scripts in a render hook is convenient because you have the specific attributes of that inputfield to hand if they are needed in the script, and the JS can be built up using the API and PHP logic and loops; I find this easier than passing API values to the JS config and doing all the dynamic/looping stuff in JS.

**Comment by ryancramerdesign** [comment]
Let me know if you find playing with the dataType helps at all. To be honest I gave up on trying to get inline scripts working through ajax requests long ago, because it seemed to be completely random as to when it worked and didn't, and generally drove me nuts. Plus inline scripts aren't really a best practice anyway, so was never too keen on them unless there was no other option. For the most part the reloaded event took over as the main way of doing this stuff. Plus I think it ends up covering a broader context overall. 

The reason Inputfield::renderReady() isn't hookable is because it's a high traffic method where making it hookable adds overhead. Since Inputfield::render() and InputfieldWrapper::renderInputfield() are already hookable, I'm cautious about adding more overhead in this sequence. There's also the fact that the method has already been in place for a long time, and appears all over not just in core modules but in 3rd party ones too... so can't really just rename it to include 3 underscores at this point. Though I understand there are those situations where renderReady() is called by something other than InputfieldWrapper::renderInputfield(), leaving you no place to hook. I'll look to see if we can add some kind of conditional hook that can activate only when necessary so that there's no overhead unless a renderReady hook is in place.

**Comment by Toutouwai** [comment]
Hi Ryan,

I did a bit more testing. I think the correct dataType of 'html' is being used so that's not the problem.

The best that I can work out is that the script elements do exist in the data variable returned by the Ajax request, and they _would be_ executed if the data was inserted in its entirety. But when a jQuery object is created from the data, which is done [here](https://github.com/processwire/processwire/blob/35df716082b779de0e53a3fcf7996403c49c9f8a/wire/modules/Fieldtype/FieldtypeRepeater/InputfieldRepeater.js#L129), then the inline scripts are not included in the object. I think this is the same thing that is warned about in the [jQuery docs for load()](http://api.jquery.com/load/#script-execution) when a selector expression is used, as I expect a jQuery object is created then also.

I'll close this issue as I don't think there can be a solution. I solved my immediate problem by listening for the reloaded event. It does necessitate adding a script to every loading of ProcessPageEdit but that isn't a big deal.
